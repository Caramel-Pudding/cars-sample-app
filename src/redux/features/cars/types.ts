interface Mileage {
  readonly number: number;
  readonly unit: string;
}

enum FuelType {
  Petrol = "Petrol",
  Diesel = "Diesel",
}

export interface Car {
  readonly stockNumber: number;
  readonly manufacturerName: string;
  readonly modelName: string;
  readonly color: string;
  readonly mileage: Mileage;
  readonly fuelType: FuelType;
  readonly pictureUrl: string;
}

interface ManufacturerCar {
  readonly name: string;
}
export interface Manufacturer {
  readonly name: string;
  readonly models: ManufacturerCar[];
}

export interface CarsState {
  readonly cars: Car[];
  readonly totalCarsCount: number;
}

export interface CarsChangePayload {
  readonly cars: Car[];
}

export interface TotalCarsCountChangePayload {
  readonly value: number;
}
