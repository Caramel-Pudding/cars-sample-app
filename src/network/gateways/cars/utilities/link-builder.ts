import { Sort } from "../../../../redux/features/sorting/types";
import { autoBasicRoute } from "../../../consts/routes";
// eslint-disable-next-line import/no-cycle
import { GetCarsRequestParams } from "../types/get-all";

export const getCarsLinkBuilder = ({
  page,
  color,
  manufacturer,
  sort = Sort.Ascending,
}: GetCarsRequestParams): string => {
  if (color && manufacturer) {
    return `${autoBasicRoute}/cars?manufacturer=${manufacturer}&color=${color}&sort=${sort}&page=${page}`;
  }
  if (color) {
    return `${autoBasicRoute}/cars?color=${color}&sort=${sort}&page=${page}`;
  }
  if (manufacturer) {
    return `${autoBasicRoute}/cars?manufacturer=${manufacturer}&sort=${sort}&page=${page}`;
  }
  return `${autoBasicRoute}/cars?sort=${sort}&page=${page}`;
};
