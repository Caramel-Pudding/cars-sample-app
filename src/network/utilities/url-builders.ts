import { Sort } from "../../redux/features/sorting/types";
import { autoBasicRoute } from "../consts/routes";
import { GetCarsRequestParams } from "../gateways/get-all";

export const buildGetAllCarsUrl = ({
  page,
  color,
  manufacturer,
  sort = Sort.Ascending,
}: GetCarsRequestParams): string => {
  if (color && manufacturer) {
    return `${autoBasicRoute}/cars?manufacturer=${manufacturer}&color=${color}&sort=${sort}&page=${page}`;
  }
  if (color) {
    return `${autoBasicRoute}/cars?color=${color}&sort=${sort}&page=${page}`;
  }
  if (manufacturer) {
    return `${autoBasicRoute}/cars?manufacturer=${manufacturer}&sort=${sort}&page=${page}`;
  }
  return `${autoBasicRoute}/cars?sort=${sort}&page=${page}`;
};

export const buildGetSignleCarUrl = (carId: string): string =>
  `${autoBasicRoute}/cars/${carId}`;

export const getColorsUrl = `${autoBasicRoute}/colors`;

export const getManufacturersUrl = `${autoBasicRoute}/manufacturers`;
