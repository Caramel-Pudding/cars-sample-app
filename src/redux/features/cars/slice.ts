import { createSlice, PayloadAction } from "@reduxjs/toolkit";

import {
  CarsState,
  CarsChangePayload,
  TotalCarsCountChangePayload,
} from "./types";

// Define the initial state using that type
export const initialState: CarsState = {
  cars: [],
  totalCarsCount: 0,
};

export const descriptorsSlice = createSlice({
  name: "Filters",
  // `createSlice` will infer the state type from the `initialState` argument
  initialState,
  /* eslint-disable no-param-reassign */
  reducers: {
    setCars: (state, action: PayloadAction<CarsChangePayload>) => {
      state.cars = action.payload.cars;
    },
    setTotalCarsCount: (
      state,
      action: PayloadAction<TotalCarsCountChangePayload>
    ) => {
      state.totalCarsCount = action.payload.value;
    },
    /* eslint-enable no-param-reassign */
  },
});

export const { setCars, setTotalCarsCount } = descriptorsSlice.actions;

export const { reducer } = descriptorsSlice;
