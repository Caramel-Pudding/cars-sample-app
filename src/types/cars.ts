export interface Car {
  readonly name: string;
}

export interface Manufacturer {
  readonly name: string;
  readonly models: Car[];
}
