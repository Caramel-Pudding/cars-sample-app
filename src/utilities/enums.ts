export const convertStringEnumToArrayOfNames = (stringEnum: {
  [stringEnum: string]: string;
}): string[] => Object.values(stringEnum);

export const convertStringEnumToArrayOfKeys = (stringEnum: {
  [stringEnum: string]: string;
}): string[] => Object.keys(stringEnum);

// TODO: Should definitely type it better
export const getEnumKeyByValue = (
  stringEnum: {
    [stringEnum: string]: string;
  },
  enumValue: string
): string =>
  Object.entries(stringEnum).find(([key, value]) => value === enumValue)![0];
