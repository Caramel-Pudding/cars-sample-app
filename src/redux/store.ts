import { configureStore, ThunkAction, Action } from "@reduxjs/toolkit";
import { reducer as filtersReducer } from "./features/filters/slice";
import { reducer as paginationReducer } from "./features/pagination/slice";
import { reducer as sortingReducer } from "./features/sorting/slice";
import { reducer as carsReducer } from "./features/cars/slice";

export const store = configureStore({
  reducer: {
    filters: filtersReducer,
    pagination: paginationReducer,
    sorting: sortingReducer,
    cars: carsReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = Readonly<ReturnType<typeof store.getState>>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
