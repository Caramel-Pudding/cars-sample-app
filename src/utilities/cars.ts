import { Car } from "../redux/features/cars/types";
import { capitalizeFirstLetter, separateNumberWithDots } from "./strings";

export const constructCarInfoString = (car: Car): string =>
  `Stock # ${car.stockNumber} - ${separateNumberWithDots(
    car.mileage.number
  )} ${car.mileage.unit.toUpperCase()} - ${
    car.fuelType
  } - ${capitalizeFirstLetter(car.color)}`;
