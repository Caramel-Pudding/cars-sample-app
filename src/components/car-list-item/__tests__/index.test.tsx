import React from "react";
import { render, cleanup } from "@testing-library/react";
import { MemoryRouter as Router } from "react-router-dom";
import { Provider } from "react-redux";

import { getCarsResponseStub } from "../../../tests/stubs";
import { CarListItem } from "..";
import { store } from "../../../redux/store";
import { constructCarInfoString } from "../../../utilities/cars";

describe("CarListItem", () => {
  const chosenCar = getCarsResponseStub.cars[0];
  const TestWrapper = (
    <Provider store={store}>
      <Router>
        <CarListItem car={chosenCar} />
      </Router>
    </Provider>
  );

  afterEach(cleanup);

  it("Renders with proper data", () => {
    const { container, getByText } = render(TestWrapper);
    const title = getByText(chosenCar.modelName);
    const carInfo = getByText(constructCarInfoString(chosenCar));
    expect(title).toBeInTheDocument();
    expect(carInfo).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
