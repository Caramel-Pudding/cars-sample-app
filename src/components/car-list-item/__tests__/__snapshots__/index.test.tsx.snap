// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarListItem Renders with proper data 1`] = `
<div>
  <article
    class="container"
  >
    <img
      alt="CLC-Klasse"
      height="85"
      src="https://auto1-js-task-api--mufasa71.repl.co/images/car.svg"
      width="100"
    />
    <section
      class="infoBlock"
    >
      <h2
        class="robotoLgBold headerTitle"
      >
        CLC-Klasse
      </h2>
      <div
        class="robotoMd"
      >
        Stock # 63428 - 100.049 KM - Diesel - Yellow
      </div>
      <a
        class="link"
        href="/details/63428"
      >
        View details
      </a>
    </section>
  </article>
</div>
`;
